python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+        *,
+        # Set the name override. If not set, uses the @activity.run decorator
+        # name. If that's not set, uses the function/method name or class name
+        # if it has __call__. If it's a lambda or nested function, a name must
+        # be provided (or it can be decorated of course).
+        name: str = None
+        # ...
+    ) -> None:
+        pass
+
+    # Cannot be called after run started. Typings and separate @overload
+    # signatures omitted for brevity.
+    def register_activities(
+        # This must be an object or list or dictionary. If an object, this only
+        # uses methods with @activity.run decorators. Dictionaries and lists
+        # don't require the decorator. If a dictionary, the key is the name.
+        # This is just a helper and is the equivalent of just calling
+        # register_activity with all applicable functions.
+        activities,
+        # ...
+    ) -> None:
+        pass
+
+    async def run() -> None:
+        pass
+```
+
+Notes:
+
+* We have left off workflow registration until a later phase
+* Stopping the worker is done via cancellation of the run task/future
+* Internally, activities will run in the same event loop `run()` is called on
+* By default will fail already-registered activities
+* Should we accept a set of activities to register in `__init__`?
+* Does exporting `temporalio.Worker` directly cause any packaging issues if we want to separate the client as its own
+  published package?
+
+### Activity Definition/Registration
+
+```python
+from temporalio import activity
+
+# This adds activity details if they are present. Maybe we should have a
+# temporalio.LoggerAdapter that works for activities _and_ workflows?
+my_activity_logger = activity.LoggerAdapter(logging.getLogger(__name__))
+
+
+def simple_activity(arg1, arg2):
+    # Uses contextvars internally, which means it'll fail in another context but
+    # we will provide the ability to set this for testing situations
+    attempts = activity.info().attempts
+    # This logger has an adapter to include custom details but still named after
+    # this module which is a common pattern
+    my_activity_logger.debug("Attempt %s", attempts)
+    # This logger, on the activity module, uses the name of the activity module.
+    # So that's why we offer either the adapter or the logger.
+    activity.logger().debug("Attempt %s", attempts)
+
+
+@activity.run(name="MyActivity")
+async def another_activity(arg1):
+    # Showing heartbeat and cancellation. In Python 3.7 we cannot give a
+    # specific error to Task/Future.cancel (only in 3.9). So activity
+    # cancellations appear like any other asyncio cancellations (which may be a
+    # good thing anyways).
+    try:
+        while True:
+            await asyncio.sleep(1)
+            await activity.heartbeat("some details")
+    except asyncio.CancelledError as exc:
+        raise RuntimeError("someone cancelled me!") from exc
+
+class MyCallableClass:
```

Comment:
Do we want activity classes? IMO they're a bit of an antipattern as they encourage attaching implicit state to something that we encourage users to think of as function invocation.

I kinda think we should push people towards making their activities take all their input as args. The counterargument here is doing things like injecting mocks, etc... but that might be nicer to get off a context?

Not super firm on this, but do want to throw it out there.

Reply:
> Do we want activity classes?

We don't care what type of callable they are using. If it's callable, we can call it.

> IMO they're a bit of an antipattern as they encourage attaching implicit state to something that we encourage users to think of as function invocation.

Not sure it is an antipattern. We often encourage users to use activity structs in Go when you want something like a client shared.

We agree that basic serializable values should be params.

Reply:
Seems unpopular so nevermind. For activities it's not a very big deal. I feel more strongly about it for workflows, though. Attaching hidden state to workflows makes it easy for different workers to end up doing different things and violate nondeterminism. So ideally there we try to make them as "pure-function-y" as possible